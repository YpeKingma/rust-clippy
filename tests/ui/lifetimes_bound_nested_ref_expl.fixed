#![warn(clippy::explicit_lifetimes_bound)]
use core::mem::MaybeUninit;

// issue 25860, missing implicit bound
pub fn lifetime_translator_1<'lfta, 'lftb: 'lfta, T>(val_a: &'lfta &'lftb T, _val_b: &'lftb T) -> &'lfta T {
    val_a
}

// helper declarations for issue 84591
trait Supertrait<'a, 'b> {
    fn convert<T: ?Sized>(x: &'a T) -> &'b T;
}

struct MyStruct;

impl<'a: 'b, 'b> Supertrait<'a, 'b> for MyStruct {
    fn convert<T: ?Sized>(x: &'a T) -> &'b T {
        x
    }
}

trait Subtrait<'a, 'b, R>: Supertrait<'a, 'b> {}

// issue 84591, missing implicit bound:
impl<'a: 'b, 'b> Subtrait<'a, 'b, &'b &'a ()> for MyStruct {}

// helper declarations for issue 100051
trait Trait1 {
    type Type1;
}

impl<T1> Trait1 for T1 {
    type Type1 = ();
}

trait Extend1<'a, 'b> {
    fn extend(self, s: &'a str) -> &'b str;
}

// issue 100051, missing implicit bound
impl<'a: 'b, 'b> Extend1<'a, 'b> for <&'b &'a () as Trait1>::Type1 {
    fn extend(self, s: &'a str) -> &'b str {
        s
    }
}

// from the httparse crate, lib.rs: bounds implied by argument and return types are the same
// missing implicit bound:
unsafe fn deinit_slice_mut<'a, 'b: 'a, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
    let s: *mut &mut [T] = s;
    let s = s as *mut &mut [MaybeUninit<T>];
    &mut *s
}

// test case for unnamed references.
// helper declarations:
struct Thing1<'a> {
    ref_u8: &'a u8,
}
struct Thing2<'a> {
    ref_u16: &'a u16,
}
// missing implicit bound
fn test_unnamed_ref<'a: 'b, 'b>(w1: &'b mut &mut Thing1<'a>, w2: &mut Thing2<'b>) -> &'a u8 {
    let _ = w2;
    w1.ref_u8
}

fn main() {}
