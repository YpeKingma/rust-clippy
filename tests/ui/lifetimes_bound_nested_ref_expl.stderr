error: missing lifetimes bound declaration: 'lftb: 'lfta
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:5:37
   |
LL | pub fn lifetime_translator_1<'lfta, 'lftb, T>(val_a: &'lfta &'lftb T, _val_b: &'lftb T) -> &'lfta T {
   |                                     ^^^^^ help: try: `'lftb: 'lfta`
   |
note: this lifetimes bound is implied here:
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:5:55
   |
LL | pub fn lifetime_translator_1<'lfta, 'lftb, T>(val_a: &'lfta &'lftb T, _val_b: &'lftb T) -> &'lfta T {
   |                                                       ^^^^^^^^^^^^
   = note: `-D clippy::explicit-lifetimes-bound` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::explicit_lifetimes_bound)]`

error: missing lifetimes bound declaration: 'a: 'b
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:25:6
   |
LL | impl<'a, 'b> Subtrait<'a, 'b, &'b &'a ()> for MyStruct {}
   |      ^^ help: try: `'a: 'b`
   |
note: this lifetimes bound is implied here:
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:25:32
   |
LL | impl<'a, 'b> Subtrait<'a, 'b, &'b &'a ()> for MyStruct {}
   |                                ^^^^^^

error: missing lifetimes bound declaration: 'a: 'b
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:41:6
   |
LL | impl<'a, 'b> Extend1<'a, 'b> for <&'b &'a () as Trait1>::Type1 {
   |      ^^ help: try: `'a: 'b`
   |
note: this lifetimes bound is implied here:
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:41:36
   |
LL | impl<'a, 'b> Extend1<'a, 'b> for <&'b &'a () as Trait1>::Type1 {
   |                                    ^^^^^^

error: missing lifetimes bound declaration: 'b: 'a
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:49:32
   |
LL | unsafe fn deinit_slice_mut<'a, 'b, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
   |                                ^^ help: try: `'b: 'a`
   |
note: this lifetimes bound is implied here:
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:49:43
   |
LL | unsafe fn deinit_slice_mut<'a, 'b, T>(s: &'a mut &'b mut [T]) -> &'a mut &'b mut [MaybeUninit<T>] {
   |                                           ^^^^^^^^^^

error: missing lifetimes bound declaration: 'a: 'b
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:64:21
   |
LL | fn test_unnamed_ref<'a, 'b>(w1: &'b mut &mut Thing1<'a>, w2: &mut Thing2<'b>) -> &'a u8 {
   |                     ^^ help: try: `'a: 'b`
   |
note: this lifetimes bound is implied here:
  --> tests/ui/lifetimes_bound_nested_ref_expl.rs:64:34
   |
LL | fn test_unnamed_ref<'a, 'b>(w1: &'b mut &mut Thing1<'a>, w2: &mut Thing2<'b>) -> &'a u8 {
   |                                  ^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

